### PR TITLE
FIX - computation of ``subdiff_distance`` in ``WeightedGroupL2`` penalty

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
## Context of the PR

This fixes the bug reported in issue #197 (thanks for the investigation @tomaszkacprzak)

## Contributions of the PR

- fix computation of `subdiff_distance`
- harden unittest

### Checks before merging PR

- ~[ ] added documentation for any new feature~
- [x] added unittests
- ~[ ] edited the [what's new](../doc/changes/whats_new.rst)~

### CODE REVIEW

File: skglm/penalties/block_separable.py [hunk 1]
```diff
@@ -324,10 +324,11 @@ def subdiff_distance(self, w, grad_ws, ws):
...
             if self.positive and np.any(w_g < 0):
                 scores[idx] = np.inf
-            if self.positive and norm_w_g == 0:
-                # distance of -norm(grad_g) to )-infty, alpha * weights[g]]
-                scores[idx] = max(0, - norm(grad_g) - self.alpha * weights[g])
-            if (not self.positive) and norm_w_g == 0:
+            elif self.positive and norm_w_g == 0:
+                # distance of -norm(neg_grad_g) to weights[g] * [-alpha, alpha]
+                neg_grad_g = np.where(grad_g < 0., grad_g, 0.)
+                scores[idx] = max(0, norm(neg_grad_g) - self.alpha * weights[g])
```
Comment (QB3) [inline skglm/penalties/block_separable.py:330]: why not using
`scores[idx] = max(0, norm(grad_g[grad_g < 0.]) - alpha * weights[g])` ?
or 
`neg_grad_g = grad_g[grad_g < 0.]) ` ?

**Review by QB3 — COMMENTED** [review]
Thank you so much for the quick fix @Badr-MOUFAD !
For future maintenance, maybe we could add details on the group Lasso subdifferential computation in the tutorial as well https://contrib.scikit-learn.org/skglm/tutorials/prox_nn_group_lasso.html#prox-nn-group-lasso

File: skglm/penalties/block_separable.py (hunk 1 above)
Comment (Badr-MOUFAD) [reply skglm/penalties/block_separable.py:330]: yes indeed @QB3, looks clear and clean

Comment (Badr-MOUFAD) [reply skglm/penalties/block_separable.py:330]: To compute the distance to the subdifferential (for group $g$)

$$
\begin{equation}
\begin{aligned}
v \in \partial \Omega & \Leftrightarrow v \in \lambda \partial \lVert \cdot \lVert_2 + \partial \delta_{\mathbb{R}_{+}^g} \\\\
 &  \Leftrightarrow v \in \lambda \mathcal{B}_2 + \mathbb{R}^{-}_g
\end{aligned}
\end{equation}
$$

hence, the distance to subdiff is the solution of this problem (Minkowski sum)

$$
\begin{equation}
\begin{aligned}
\min_{u, n} & \ \lVert u + n - v \rVert^2 \\
\mathrm{st.} & \ u \in \lambda \\mathcal{B}_2, n \in \\mathbb{R}_g^{-}
\end{aligned}
\end{equation}
$$

The implemented formula solves this problem by solving first for $n \in \\mathbb{R}_g^{-}$, then for $u \in \lambda \\mathcal{B}_2$
Intuitively, it seems right to do that for this case, but I can't prove it rigorously

Comment (QB3) [reply skglm/penalties/block_separable.py:330]: Maybe we can extend [this proof](https://math.stackexchange.com/questions/2755045/how-is-the-subdifferential-of-the-l-2-norm-at-x-0-the-polar-of-the-unit-ball).
I would vote for merging the fix quickly, and then doing a PR to detail the subdifferential computation.

Comment (mathurinm) [reply skglm/penalties/block_separable.py:330]: @Badr-MOUFAD can you include the derivation in https://contrib.scikit-learn.org/skglm/tutorials/prox_nn_group_lasso.html#prox-nn-group-lasso ? Since this is not "trivial"

You're looking at a projection problem; it's clear that for the negative entries, their value in the solution is the same; then you're left with the projection of the positive entries on the ball + R_-, wich is clearly the projection on the ball too (as projecting onto R- would imply changing the sign, which is suboptimal)

Comment (mathurinm) [reply skglm/penalties/block_separable.py:330]: >  I would vote for merging the fix quickly, and then doing a PR to detail the subdifferential computation.

I disagree, we tend to not finish stuff that is "left for later"
The proof is not complex (we can even just give the solution)

**Comment by mathurinm** [comment]
@Badr-MOUFAD @QB3 I'm done with the pass on docs, make any changes you want and merge

**Review by mathurinm — APPROVED** [review]

**Comment by Badr-MOUFAD** [comment]
@QB3 a glance at the derivation from you is a plus

File: doc/tutorials/prox_nn_group_lasso.rst [hunk 2]
```diff
@@ -135,5 +134,65 @@ and thus, combined with Equations :eq:`prox_projection_nn_Sc` and :eq:`prox_proj
...
+    \lambda \partial || \cdot ||_2 + \partial \iota_{x \geq 0} = \lambda \mathcal{B}_2 + \mathbb{R}_-^g
+    ,
+
+where :math:`\mathcal{B}_2` is the unit ball.
+
+Therefore, the distance to the subdifferential writes
+
+.. math::
+
+    D(v) = \min_{u \in \lambda \mathcal{B}_2, n \in \mathbb{R}_{-}^g} \ || u + n - v ||
```
Comment (mathurinm) [inline doc/tutorials/prox_nn_group_lasso.rst:182]: @Badr-MOUFAD this hides the problem, because why is the solution for a negative v, n = 0? 

I'm subjective but I find it way clearer the way it was before, which avoids reference to an external source

Comment (Badr-MOUFAD) [reply doc/tutorials/prox_nn_group_lasso.rst:182]: > @Badr-MOUFAD this hides the problem, because why is the solution for a negative v, n = 0?

if v is negative, one can set n = v and u = 0 to get a distance equal zero (zero is a lower bound on the problem)

> I'm subjective but I find it way clearer the way it was before, which avoids reference to an external source

The external source is a stack overflow thread to justify the computation of the min regardless of the order.

I believe the benefit of having formulas is to easily and efficiently look up the derivation later.
thought, I'm +1 for adding an explanation, that way we can have both worlds.

Comment (QB3) [reply doc/tutorials/prox_nn_group_lasso.rst:182]: > @QB3 a glance at the derivation from you is a plus

I also think Mathurin's version was clearer.  I would also try to be more self-contained (e.g. we currently talk about the distance, without always clearly stating that this is the distance of the gradient to the subdiff), but writing something clear and self-contained might take more time.

TLDR: I am happy to merge Mathurin's version and iterate later on the group lasso doc

**Review by mathurinm — APPROVED** [review]
